### PR TITLE
fix(sdk.v2): Component is invalid if `packages_to_install` empty and `install_kfp_package=False`

### PR DESCRIPTION
**Pull request by judahrand**
**Description of your changes:**
* Add test for empty `packages_to_install_command`
* Don't return `pip install` with no packages
    
   If an empty list is provided no command should be returned to avoid a
   broken command eg. `pip install      `

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

**Comment by judahrand** [comment]
@googlebot I signed it!

File: sdk/python/kfp/components_tests/test_python_op.py [hunk 1]
```diff
@@ -1158,6 +1158,12 @@ def test_packages_to_install_feature(self):
             self.helper_test_component_using_local_call(
                 task_factory2, arguments={}, expected_output_values={})
 
+        task_factory3 = comp.func_to_container_op(
+            dummy_in_0_out_0, packages_to_install=[], install_kfp_package=False)
```
[the remark this answers is not in the document]
Comment (judahrand) [reply sdk/python/kfp/components_tests/test_python_op.py:1162]: I verified that the test I wrote fails without my changes and passes with them?

File: sdk/python/kfp/v2/components/component_factory.py [hunk 2]
```diff
@@ -35,7 +35,7 @@ def _python_function_name_to_component_name(name):
 def _get_packages_to_install_command(
         package_list: Optional[List[str]] = None) -> List[str]:
     result = []
-    if package_list is not None:
```
[the remark this answers is not in the document]
Comment (judahrand) [reply sdk/python/kfp/v2/components/component_factory.py:38]: I'd argue that it could also be a user explicitly expressing no packages. It's clearer in code than `None` or leaving the default.

In fact I did exactly that which is what prompted the PR.

[the remark this answers is not in the document]
Comment (judahrand) [reply sdk/python/kfp/v2/components/component_factory.py:38]: That's not true, see: https://github.com/kubeflow/pipelines/blob/d83933dafd7ae84659c9dff7d170b8655b81df82/sdk/python/kfp/v2/components/component_factory.py#L154

In the code the `None` is replaced with an empty list anyway. So leaving the default `None` will result in the empty list.

[the remark this answers is not in the document]
Comment (judahrand) [reply sdk/python/kfp/v2/components/component_factory.py:38]: That's not true, see: https://github.com/kubeflow/pipelines/blob/d83933dafd7ae84659c9dff7d170b8655b81df82/sdk/python/kfp/v2/components/component_factory.py#L154

In the code the `None` is replaced with an empty list anyway. So leaving the default `None` will result in the empty list.

File: sdk/python/kfp/components_tests/test_python_op.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (judahrand) [reply sdk/python/kfp/components_tests/test_python_op.py:1162]: `func_to_container_op` calls `_func_to_component_spec` which calls `_get_packages_to_install_command`

https://github.com/kubeflow/pipelines/blob/2f19a26ffd205c7a3384c4c250f5dbb5150a7c4d/sdk/python/kfp/components/_python_op.py#L928

You are correct about the `install_kfp_package` arg, however.

[the remark this answers is not in the document]
Comment (judahrand) [reply sdk/python/kfp/components_tests/test_python_op.py:1162]: You're right, let me remove the commit which adds this test. Do you think this is worth adding a test for?
It certainly caused me problems...

File: sdk/python/kfp/v2/components/component_factory.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/v2/components/component_factory.py:38]: I see, maybe an alternative change it to move `packages_to_install = packages_to_install or []` on line 154 to line 158:

https://github.com/kubeflow/pipelines/blob/2f19a26ffd205c7a3384c4c250f5dbb5150a7c4d/sdk/python/kfp/v2/components/component_factory.py#L154-L158

WDYT?

[the remark this answers is not in the document]
Comment (judahrand) [reply sdk/python/kfp/v2/components/component_factory.py:38]: Personally I think I prefer the current change and would maybe also remove the `Optional` from `_get_packages_to_install_command`.

Imo `_get_packages_to_install_command` should have enough logic (and does) to identify when there are no packages (empty list). It accepting a 'None` type too just muddies the water in my mind. I'm not sure I really understand the advantage of it taking `None`

[the remark this answers is not in the document]
Comment (neuromage) [reply sdk/python/kfp/v2/components/component_factory.py:38]: > It accepting a None type too just muddies the water in my mind. I'm not sure I really understand the advantage of it taking None

Please don't remove the `Optional`. It's not a good idea to use mutable objects as default arguments:
https://google.github.io/styleguide/pyguide.html#212-default-argument-values

[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/v2/components/component_factory.py:38]: I personally prefer the alternative change I mentioned above, in which way `packages_to_install == None` indicates system default, whereas `packages_to_install == []` must be from user input. 

That said, I'm fine with the current change as well.

**Review by chensun — COMMENTED** [review]
Hi @judahrand, can you please add an entry for this PR in the SDK release note: https://github.com/kubeflow/pipelines/blob/master/sdk/RELEASE.md

File: sdk/python/kfp/v2/components/component_factory.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (judahrand) [reply sdk/python/kfp/v2/components/component_factory.py:38]: @neuromage That's not what I was suggesting at all! I'm fully aware that mutable default args are a very bad idea.

My suggestion was simply to remove the `Optional` type hint on `_get_packages_to_install_command` to make it clear that a list is expected.

File: sdk/RELEASE.md [hunk 3]
```diff
@@ -11,6 +11,7 @@
 ## Deprecations
 
 ## Bug Fixes and Other Changes
+* Fix component decorator could result in invalid component if `install_kfp_package=False`. [\#6527](https://github.com/kubeflow/pipelines/pull/6527))
```
Comment (chensun) [inline sdk/RELEASE.md:14]: nit: there's extra `)` at the end. I'll fix it later when doing the release.

**Review by chensun — COMMENTED** [review]
/lgtm
/approve

Thanks!

File: sdk/python/kfp/v2/components/component_factory.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chensun) [reply sdk/python/kfp/v2/components/component_factory.py:38]: > It's a private function (as much as such a thing exists in Python) which is only used internally so there's not really a good reason for the Optional in my mind. And it already doesn't have a default value.

`Optional[...]` is just a shorthand for `Union[None, ...]`, it has nothing to do with default value. With the alternative change I commented above, we would get `None`--the default value from the public method--passed all the way through to this point. So `Optional` would make sense.
